File: c/onp.c

/**
 * The algorithm implemented here is based on "An O(NP) Sequence Comparison Algorithm"
 * by described by Sun Wu, Udi Manber and Gene Myers
 */

#include "onp.h"

static uint snake (onp_diff_t *diff, int k, int p, int pp, int offset);
static uint snake (onp_diff_t *diff, int k, int p, int pp, int offset) {
  int y = onp_max(p, pp);
  int x = y - k;
  while (x < diff->m && y < diff->n && diff->a[x] == diff->b[y]) {
    ++x;
    ++y;
  }
  return y;
}

onp_diff_t *onp_alloc_diff (onp_sequence_t *a, onp_sequence_t *b) {
  onp_diff_t *diff = (onp_diff_t *)malloc(sizeof(onp_diff_t));
  diff->a = a;
  diff->b = b;
  diff->m = onp_seq_len(a);
  diff->n = onp_seq_len(b);
  if (diff->m > diff->n) {
    onp_swap(onp_sequence_t*, seq_tmp, a, b);
    onp_swap(int, int_tmp, diff->m, diff->n);
  }
  diff->editdis = 0;
  return diff;
}

void onp_free_diff (onp_diff_t *diff) {
  free(diff->path);
  free(diff);
  diff = NULL;
}

void onp_compose (onp_diff_t *diff) {
  int  p      = -1;
  int  delta  = diff->n - diff->m;
  uint offset = diff->m + 1;
  uint size   = diff->m + diff->n + 3;
  int *fp     = (int *)malloc(sizeof(int) * size);
  diff->path  = (int *)malloc(sizeof(int) * size);
  for (int i=0;i<size;++i) {
    fp[i] = diff->path[i] = -1;
  }
  do {
    ++p;
    for (int k=-p;k<=delta-1;++k) {
      fp[k+offset] = snake(diff, k, fp[k-1+offset]+1, fp[k+1+offset], offset);
    }
    for (int k=delta+p;k>=delta+1;--k) {
      fp[k+offset] = snake(diff, k, fp[k-1+offset]+1, fp[k+1+offset], offset);
    }
    fp[delta+offset] = snake(diff, delta, fp[delta-1+offset]+1,
                             fp[delta+1+offset], offset);
  } while (fp[delta+offset] != diff->n);
  diff->editdis = delta + 2 * p;
  free(fp);
}

File: go/onp/onp.go
/**
 * The algorithm implemented here is based on "An O(NP) Sequence Comparison Algorithm"
 * by described by Sun Wu, Udi Manber and Gene Myers
 */

package onp

import (
	"math"
)

func max(x, y int) int {
	return int(math.Max(float64(x), float64(y)))
}

type Diff struct {
	str1, str2 string
	m, n       int
}

func NewDiff(str1, str2 string) *Diff {
	if len(str1) > len(str2) {
		return &Diff{str2, str1, len(str2), len(str1)}
	}
	return &Diff{str1, str2, len(str1), len(str2)}
}

func (diff *Diff) snake(k, p, pp int) int {

	y := max(p, pp)
	x := y - k

	for x < diff.m && y < diff.n && diff.str1[x] == diff.str2[y] {
		x += 1
		y += 1
	}

	return y
}

const NOT_SOLVED = -1

func (diff *Diff) EditDistance() int {

	offset := diff.m + 1
	delta  := diff.n - diff.m
	size   := diff.m + diff.n + 3
	fp     := make([]int, size)

	for i := range fp {
		fp[i] = -1
	}

	for p := 0; ; p++ {

		done := make(chan bool)
		go func() {
			for k := -p; k <= delta-1; k++ {
				fp[k+offset] = diff.snake(k, fp[k-1+offset]+1, fp[k+1+offset])
			}

			done <- true
		}()

		go func() {
			for k := delta + p; k >= delta+1; k-- {
				fp[k+offset] = diff.snake(k, fp[k-1+offset]+1, fp[k+1+offset])
			}
			done <- true
		}()

		for i := 0; i < 2; i++ {
			<-done
		}
		fp[delta+offset] = diff.snake(delta, fp[delta-1+offset]+1, fp[delta+1+offset])

		if fp[delta+offset] >= diff.n {
			return delta + 2*p
		}
	}

	return NOT_SOLVED
}
